compiler/rustc_hir_typeck/src/callee.rs

```diff
@@ -137,19 +136,17 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         output
     }
 
-    /// Functions of type `extern "custom" fn(/* ... */)` cannot be called using `ExprKind::Call`.
-    ///
-    /// These functions have a calling convention that is unknown to rust, hence it cannot generate
-    /// code for the call. The only way to execute such a function is via inline assembly.
-    fn check_call_custom_abi(&self, callee_ty: Ty<'tcx>, span: Span) {
+    fn check_call_abi(&self, callee_ty: Ty<'tcx>, span: Span) {
         let abi = match callee_ty.kind() {
             ty::FnDef(def_id, _) => self.tcx.fn_sig(def_id).skip_binder().skip_binder().abi,
             ty::FnPtr(_, header) => header.abi,
             _ => return,
         };
 
-        if let ExternAbi::Custom = abi {
-            self.tcx.dcx().emit_err(errors::AbiCustomCall { span });
+        // Some ABIs cannot be called from rust, either because rust does not know how to generate
+        // code for the call, or because a call does not semantically make sense.
+        if !abi.can_be_called_with_call_expr() {
+            self.tcx.dcx().emit_err(crate::errors::AbiCannotBeCalled { span, abi: abi.as_str() });
```

Comment:
In the case of `interrupt`, we also do not want to reach any `iret` instructions.

Reply:
Right, I made a note of that in a comment.